### PR TITLE
Fix the `.get()` TypeScript return type

### LINKED ISSUE #116: TypeScript types should take into account defaults
https://github.com/sindresorhus/electron-store/blob/966a9c736318c6c51083de08e5eb5d1276e00047/index.test-d.ts#L47 should not be `number | undefined` as it already has a default value defined in the constructor: https://github.com/sindresorhus/electron-store/blob/966a9c736318c6c51083de08e5eb5d1276e00047/index.test-d.ts#L40-L43

### PR DESCRIPTION
**Pull request by SnosMe**
Closes #116

### CODE REVIEW

File: source/index.ts [hunk 1]
```diff
@@ -165,9 +165,11 @@ class Conf<T extends Record<string, any> = Record<string, unknown>> implements I
 	@param key - The key of the item to get.
 	@param defaultValue - The default value if the item does not exist.
 	*/
-	get<Key extends keyof T>(key: Key | string): T[Key] | undefined;
-	get<Key extends keyof T, Default = T[Key]>(key: Key | string, defaultValue: Default): T[Key] | Default;
-	get<Key extends keyof T, Default = T[Key]>(key: Key | string, defaultValue?: Default): Default | undefined {
+	get<Key extends keyof T>(key: Key): T[Key];
+	get<T = unknown>(key: string): T;
+	get<Key extends keyof T, Default = Required<T>[Key]>(key: Key, defaultValue: Default): Required<T>[Key] | Default;
```
Comment (sindresorhus) [inline source/index.ts:170]: The return value and the default value should always be the same type.

File: source/index.ts [hunk 2]
```diff
@@ -165,9 +165,11 @@ class Conf<T extends Record<string, any> = Record<string, unknown>> implements I
...
+	get<T = unknown, Default = T>(key: string, defaultValue: Default): T | Default;
+	get(key: string, defaultValue?: any): any {
```
Comment (sindresorhus) [inline source/index.ts:172]: Don't use `any`

File: test/index.test-d.ts [hunk 3]
```diff
@@ -127,6 +127,8 @@ expectType<boolean>(config.get('isRainbow', false));
 
 expectType<string | undefined>(config.get('unicorn'));
 expectType<string>(config.get('unicorn', 'rainbow'));
+// @ts-expect-error
+expectType<number>(config.get('unicorn', 1));
```
Comment (sindresorhus) [inline test/index.test-d.ts:131]: `tsd` has a `expectError` assertion.

File: source/index.ts [hunk 4]
```diff
@@ -165,9 +165,10 @@ class Conf<T extends Record<string, any> = Record<string, unknown>> implements I
 	@param key - The key of the item to get.
 	@param defaultValue - The default value if the item does not exist.
 	*/
-	get<Key extends keyof T>(key: Key | string): T[Key] | undefined;
-	get<Key extends keyof T, Default = T[Key]>(key: Key | string, defaultValue: Default): T[Key] | Default;
-	get<Key extends keyof T, Default = T[Key]>(key: Key | string, defaultValue?: Default): Default | undefined {
+	get<Key extends keyof T>(key: Key): T[Key];
+	get<Key extends keyof T>(key: Key, defaultValue: Required<T>[Key]): Required<T>[Key];
+	get<Key extends string, TValue = unknown>(key: Key extends keyof T ? never : Key, defaultValue?: TValue): TValue;
```
Comment (sindresorhus) [inline source/index.ts:170]:
```suggestion
	get<Key extends string, Value = unknown>(key: Key extends keyof T ? never : Key, defaultValue?: Value): Value;
```

Comment (sindresorhus) [reply source/index.ts:170]: Could you add a short code comment about the purpose of this overload? It's not immediately clear.

Comment (SnosMe) [reply source/index.ts]: Its purpose is commit name: using incorect type for default value should not fallthrough to overload with unknown type.

If you want to put it in source I'm not sure how to do it elegantly 😄

Comment (sindresorhus) [reply source/index.ts]: Just write what you wrote above, just in a code comment.

File: test/index.test-d.ts [hunk 5]
```diff
@@ -67,7 +67,7 @@ conf.set({
 	}
 });
 
-expectType<string | undefined>(conf.get('foo'));
+expectType<string>(conf.get('foo'));
```
Comment (Superjo149) [inline test/index.test-d.ts:70]: The second one was indeed incorrect. But Is it correct to assume that a value will never be undefined when we do not pass a default? I think we should still make it possible to be undefined to encourage people to handle an undefined state.

Comment (SnosMe) [reply test/index.test-d.ts:70]: Yes, value will never be undefined unless developer make a bug such in this test, type `UnicornFoo` is source of truth.
```ts
type UnicornFoo = {
	foo: string;
};
const conf = new Conf<UnicornFoo>({accessPropertiesByDotNotation: true});
```
Maybe we should validate `default` option, but it is out of scope of this PR.
Like this
```ts
	defaults?: Readonly<{
		[P in keyof T]: Exclude<T[P], undefined>
	}>;
```

Comment (sindresorhus) [reply test/index.test-d.ts:70]: > Maybe we should validate default option, but it is out of scope of this PR.

That would be nice, yes, but I agree, it should be a separate PR.

**Comment by sindresorhus** [comment]
Thanks for fixing this, @SnosMe 🙌🏻

**Pull request merged**